### ISSUE TITLE
Accessing an array from multiple workers

### ISSUE THREAD

**Issue by seme1** [TopCloud/socketcluster#44]
Currently, I have one single worker running with a global array "ActiveUsers = []". Inside that array, I store a list of the ids for all logged in users that subscribed to some private channels. 

My question is: when spawning two workers say worker1 and worker2, what is the best way for accessing worker1.ActiveUsers from inside worker2 ?

**Comment by seme1** [comment]
I've been thinking about the best setup for storing the IDs of all logged in users when having two (or more) worker processes. Perhaps it's best to store a portion of the authenticated users in Worker1 and another portion in Worker2 instead of using a global array shared between both workers that would contain all users IDs.

In this case my question is, what is the best way for transferring day from one worker to all the rest of the running workers ?? perhaps a 'private' subscribe/publish scheme can be applied so whenever a new worker is launched, it would subscribe to the 'private' channel. In this case if it requires to check whether is a userID is already registered in one of other workers arrays, it would simply publish a question to the rest of the workers in a specific format ??

**Comment by MegaGM** [comment]
When SC was ~0.9.82, I was just using Redis pub/sub. It's straightforward and quite efficient way.

Now I've planned to make another app with SC power, and I wonder is there a more nifty way to implement such functionality?
@jondubois

**Comment by MegaGM** [comment]
Btw it's fairly popular question, mb it will be fine to have an example in docs ![](http://smayly.ru/gallery/anime/RedFox/16.gif)

**Comment by seme1** [comment]
I am trying to figure out how to implement this with Redis. What I'm trying to achieve is a way for different workers to interchange data that should only be available while the workers up and running. It should also be flexible so that if a worker is added or suddenly removed, a worker should still provide the correct results by communicating with only the other "running" workers.

One difficulty I'm facing when using the pub/sub model for solving the above problem is how could worker1 know how many other workers are available and still in a "running" state in real time ? 

To clarify this, let's assume that all workers subscribe to a "private" channel once they are initiated. Let's say also assume that there are 3 workers : worker1, worker2 and worker3. If worker1 needs to retrieve data from worker2 and worker3, it would have to publish to a request through the "private" channel and wait for a response from all available workers. The responses from worker2 and worker3 are expected to be separate but in the same "private" channel. The question is : for how long should worker1 wait for a response from all other workers before deciding that all other workers have already responded ? Since worker1 does NOT know how many other workers are still available, I'm guessing this would be very trivial to solve.

**Comment by jondubois** [comment]
@seme1 You should never try to make each worker share data with all other workers... You don't want to be in a situation where each worker knows about the state of the entire system - That kills scalability. Also you shouldn't store important state in your workers.

You can use something like Redis, MongoDB, Cassandra or SC's own scServer.global object (http://socketcluster.io/#!/docs/api-global) to share data between your workers - The solution you choose depends on how critical that data is. For temporary data like the list of activeUsers, it should be OK to use the scServer.global object. You can specify a mapping function (using global.setMapper()) in case you want to spread your data evenly across multiple stores (for scalability) - Read the docs carefully if you want to understand how this works (it's tricky). More docs coming soon.

With pub/sub, it's important to note that you often don't need to know about who is interested in some data in order to deliver that data to them. Channels can make things much easier.

You can create channels for any kind of data - You could let each user have a unique channel to represent his 'online_friends' for example... Whenever one of his friends comes online, you can publish data to his channel to let the user know - In this scenario, the backend only has to track friends online status on a user-by-user basis - So no need to keep one massive array of all users who are online.

@MegaGM You can subscribe to pub/sub channels on the server (see scServer.global http://socketcluster.io/#!/docs/api-global) - To keep the channel private, you need to add a MIDDLEWARE_SUBSCRIBE middleware to block all clients from subscribing to these private channels - That way only server-side processes will be able to subscribe to them. See addMiddleware method in API docs: http://socketcluster.io/#!/docs/api-scserver

Maybe this is something that needs to be simplified? I'm happy to hear suggestions.
Should private server-only channels become a new concept? Or maybe this usage pattern (using middleware) should be explained better in the docs? - I think Middleware should be a central part of SC (just as they are in Express).

I'm putting this in my TODOs list. You shouldn't have to use Redis to share private data between workers :p Thanks for pointing this out.

**Comment by MegaGM** [comment]
> how could worker1 know how many other workers are available and still in a "running" state in real time ?
> how long should worker1 wait for a response from all other workers before deciding that all other workers have already responded ?

Nooo, it's a bad design. Workers should **not** know about other workers at all.
Each worker should be totally independent.
All workers should handle only data, that they able to reach **inside themselves**.
Workers are your ~~slaves~~ workers! You should never ask them something.
You should only give them orders and directions.

---

#### RIGHT WAY
Give directions to workers:
1. When you encounter a message "kick all, who under 18 years old" from the client-side => say to Redis "kick all under 18"
2. When you encounter a message "kick all under 18" from Redis => kick all users that you know, that are less than 18 years old
And each worker should check whether inside of it there is such users, then kick them if there is.

---

#### WRONG WAY
Give directions to workers:
1. If somebody asks you "are there another users under 18?" => search such users inside yourself, and give back an answer
2. When you encounter a message "kick all, who under 18 years old" from the client-side => search such users inside yourself, then additionally ask all other workers via Redis, "are there another users under 18?", and wait "SOME TIME" for answers
3. Kick all users that you know, that are less than 18 years old, and kick all users that other workers told you
Nope, nope, no. ^^^^^^^^ It's a bad design.

---

If in your application there is some type of data that must be available in all workers - go ahead and store it somewhere else, really.

1. Do you need persistent sessions in your app? Store them in a database (Redis? :3)

2. Do you need messages in your app? Store them in a database (Redis? :3)

3. You want to notify users when a new message comes?

  1. A worker got a message from the client-side, and put it in a database.

  2. The worker talks a special phrase with necessary data to Redis channel "hey workers, there is a new message, please notify to all".

  3. All workers always are waiting for special phrases and always can handle it. Even if they were born just right now, and they are interacting with no clients yet, it shouldn't matter.

  4. Last but not least, that "array of users"(probably not an array but an object) contains not users, but **socket sessions**. Of course workers should add them/remove them properly. And if you need more data in future to filter who can receive messages, put the data inside socket session at this step. 

---

Let's say, your client opened 1 or 3 or 30 tabs in a browser. If I remember correct, them all should be attached to the same worker and all that 30 sockets objects should share the same **socket session**. If another client open another browser then all those tabs also should share another one session, it doesn't matter how many tabs opened there.
So, what to do when the special phrase has come into a worker? Loop thru all collected socket session and emit another special phrase "hey client, a new message has been received". This message will be delivered to all of those 30 tabs.

Let's say you want to implement a feature like "mark as read". When you hit a button "mark as read" 3(i,ii,iii,iv) are repeated.

Or if you make really big app, each of tabs, when gets a message, has to store info about the message into LocalStorage, and listen to ``` window.addEventListener('storage', ... ) ```, and when you press the button, instead of  31 interactions (1: "hey server, mark this as read", 2-31: "hey client, this was marked as read by you somewhere else") there will be only 1.
And so on... 
---
Oh, sorry :\D Now, finally the time to give an answer to your question ![](http://smayly.ru/gallery/anime/RedFox/2.gif)
1. Store any data that should be reached in *all workers* in a database.
2. Populate often used necessarily data from the database to socket session (on 'connection' event). 
3. If there is necessarily, but rarely used data - query it from the database, when it needs, inside one worker, and pass to Redis for all workers, if you need the data in all workers :\) 

I hope you get the point.

---

And last one for today. I didn't work with SC bunch of time, so, maybe now there is a better way to organize workflow ;D
Thank you for reading, but it all needs to be confirmed by @jondubois ![](http://smayly.ru/gallery/anime/RedFox/3.gif)

**Comment by MegaGM** [comment]
OMG, the page was opened ~5 hours, while I was working, and when I've come back to this tab, to finish the reply, I didn't noticed Jonathan's message. I'm so ashamed :'(

**Comment by jondubois** [comment]
@MegaGM Lol. Thanks for the LONG, enthusiastic response! Record breaking! That's what this community is all about :)))

You really nailed the root of the issue with the 'Right way' vs 'Wrong way' - Can I use this idea for a new documentation page to add to the website?

The only thing I would add is that you don't necessarily **need** to use Redis's pub/sub feature to talk between workers (for private channels) (see my post above) - SC already shares channels between workers automatically and you can use middleware to make sure that they are not accessible by clients. Of course SC has no strict rules - It really depends on your specific environment and what's easier for you :)

You do need a database of sorts if you want to store persistent data. But since SC can do all the pub/sub stuff on its own, you are not limited to just Redis - MongoDB, Cassandra, or even plain old MySQL or PostgreSQL is fine!

**Comment by seme1** [comment]
Thanks to @MegaGM  and @jondubois  for the very informative responses and comments. I learned a lot especially from the examples provided by @MegaGM . Thanks to you all. I also think it will be very useful if the comments provided by @MegaGM are added to the website Documentation secion. 

However, I'm still puzzled on how to solve the following problem while still making each worker run independently from the rest of the workers. Here is what I have:

Each worker has an `activeUsers` object which basically stores the IDs of the currently logged in users in this worker. It also links each user ID to the session(s) ID of the user. A logged in user may have one or more sessions.

What I'm trying to achieve is to update a backend MySQL database with the number of LIVE socket sessions each user has. Initially, I implemented this via having `UserSessionsNumber` counter in the database that is initialized to zero for each user. When a user is authenticated via a websocket, a connection is made to MySQL to increase the `UserSessionsNumber` by 1. When a user session ends because the user either logged out or got disconnected, the `UserSessionsNumber` counter is decremented by 1.

The above solution works nicely and has the advantage of that every worker does NOT need to know or interact with the rest of the workers for computing the number of live sessions a user has/have. HOWEVER, when a worker crashes or suddenly stops, the `sessionEnd` event for each user is never called. Thus, the `UserSessionsNumber` is NOT decremented leading to the values stored in the DB to be INCORRECT. 

I spent too much time trying to figure out the best strategy for avoiding the problem that may arise when one of the workers crashes. The only way I could think of for avoiding this problem is by iterating through the `activeUsers` variable to compute how many active sessions a user has every time he/she signs in or sign out. If there is more than one worker running, then this means that the worker would somehow have to know how many active sessions this user has in the other running workers so that it can update the MySQL database with the correct value.

I understand now that the above is a bad design. However, if there is any other strategy you suggest for computing the total number of live sessions each user has, please let me know.

**Comment by seme1** [comment]
I just thought about another idea. Perhaps I could have each user session `subscribe` to a private user Channel, perhaps identified with the user_Id. SocketCluster should (based on my understanding ) span this private channel across all running workers. In this way, when a session is created for a user in worker1 and another session is created for the same user in worker2, both user sessions should subscribe to the channel `private_userId`. This way, when a worker crashes, all of the subscriptions initiated by this worker would also be deleted.

The question I really need an answer for to make the above work is: how to find the total number of subscriptions made to any channel ?

**Comment by jondubois** [comment]
@seme1 Yes, I think what you're described in your last post is the best flow!
You shouldn't try to make an effort to store all this activeUsers information to map usernames to session ids - It's a lot of work to maintain this array across multiple workers.

In any case; regarding sockets from the same session going to different workers - That is not possible as of v1.2.5 - SC does sticky balancing which forces all sockets opened in the same browser (across multiple tabs for example) to be bound to the same worker.

I feel that since pub/sub channels were introduced in SocketCluster v1, the idea of a session has become somewhat redundant and people are often tempted to 'keep track of them' which is probably more of an anti-pattern which we want to avoid.

I am working on SocketCluster v2 right now and we are getting rid of the session object in favor of just using channels based on user names because it's so much more powerful.

One of the things which SC emphasizes is that if your code doesn't scale across multiple processes (vertically), it also won't scale horizontally across multiple machines either (assuming your product gets to that stage). Conversely, you will find that if your code can scale linearly across multiple processes on a single, big machine using SC (and is capable of recovering from worker crashes), that same code should also be able to scale across multiple machines without requiring any major changes. The two types of scalability require the same programming design patterns so SC encourages you to think about these things early in the project.

**Comment by seme1** [comment]
Thanks again for your response. The idea of having sessions replaced by usernames (or userIds) is really great. I'm eagerly waiting for it to come out so that I can implement it in my system. Can you give us a hint on its ETA ? I may put my current development on this aspect of the project on hold until it comes out since I'll most probably have to change the design of many things on the client side/server side, especially the authentication part. For me sessions do not have much value as the app will be cordova-based and will run only once on each device. This also means that the same user may login from different devices.

Meanwhile, I came across [this post](https://coderwall.com/p/ekrcyw/socket-io-managing-single-user-multi-connections) which is very similar to the last idea I had for tracking user IDs through pub/sub channels. The question I had is how to get the total number of subscriptions made to any particular channel ? 

Another way for asking the same question: when examining the following io.socket code (which is taken from the above link):
`var counter = io.sockets.clients(socket.room).length;`
How to get the value of counter by using SC instead of io.sockets  ?

**Comment by MegaGM** [comment]
@jondubois wrote:
>  'Right way' vs 'Wrong way' - Can I use this idea for a new documentation page to add to the website?

Of course! I guess this kind of comparison will help people to understand all better.


---


@seme1 wrote:
> I spent too much time trying to figure out the best strategy for avoiding the problem that may arise when one of the workers crashes.

Okay, I'll try to help you.
If I understand you correct, in your application you wanna grant users an ability to be LoggedIn from several devices simultaneously.

##### We have already something like that:
    1. User entity (entity or model or table in db, doesn't matter how to call it, right?)
    2. Session entity
    3. One-to-many relation between User->Session
    4. A bunch of pivot tables, because you use MySQL. It's okay.

Okay, now we should create another table, to track sessions' online.

| online |
| ----- |
| sessionId |
| userId |
| workerPID |



Go ahead and imagine what our workers should do.

##### Whenever a user does "SignIn":

    1. Change something in User and Session entities (e.g ` User.online = 1 ` etc)
    1. Add relation between User and Session
    3. Add a new record into ` online `

Now we able to perform queries like ` get all online users ` or ` get all online sessions ` or even ` get all online sessions by workerPID `. Aha, sounds good but we still have a big problem.

@seme1 wrote:
> the problem that may arise when one of the workers crashes.

For finding out a solution, let's imagine what happens whenever a worker does "oh noo I'm died.."?
Right! A new worker is borning. So, it's the right moment to us to clear ` online ` table from users, that actually is not online anymore, cuz previous worker has died. What our slave should do?

##### Whenever a worker is died and/or a new one is started:

    1. Ask a mysterious Master process "Hello, my Master! Give me, please, PIDs of all workers"
    2. Select all rows in the ` online ` table, that have unknown ` workerPID `. (let's call it ` dirtyOnline `)
    3. Do some additional logic, you know what you want. Just for instance: for each user in ` dirtyOnline ` check whether he has another online sessions, if no then change ` User.online = 0 ` etc
    4. Finally delete all ` dirtyOnline ` rows

This approach will also clear entire ` online ` table when you perform cold start your application.
And ` get all online session where userId X ` it's probably a good answer to your question:
> What I'm trying to achieve is to update a backend MySQL database with the number of LIVE socket sessions each user has.

Okay, now it's sounds good. Good enough? Actually **no**, I never ever tested it, and probably never will.

---

But as I see you wanna change your application. And it's good decision. Let's say now we create two SC channels ` online_users ` and ` online_anons `.
What's now?


##### Whenever a user does "SignIn":

    1. Unsubscribe user from ` online_anons `
    2. Subscribe user for ` online_users `
    3. Perform additional logic, you know, something like "Add relation between User and Session"

##### Whenever a worker is died and/or a new one is started:

    0. **Nothing**

But now we have even more questions. And those questions for @jondubois 
I'll make another issue with list of those questions. @seme1 please be free to add there something.

**Comment by MegaGM** [comment]
@jondubois wrote:
>  'Right way' vs 'Wrong way' - Can I use this idea for a new documentation page to add to the website?

Of course! I guess this kind of comparison will help people to understand all better.


---


@seme1 wrote:
> I spent too much time trying to figure out the best strategy for avoiding the problem that may arise when one of the workers crashes.

Okay, I'll try to help you.
If I understand you correct, in your application you wanna grant users an ability to be LoggedIn from several devices simultaneously.

##### We have already something like that:
    1. User entity (entity or model or table in db, doesn't matter how to call it, right?)
    2. Session entity
    3. One-to-many relation between User->Session
    4. A bunch of pivot tables, because you use MySQL. It's okay.

Okay, now we should create another table, to track sessions' online.

| online |
| ----- |
| sessionId |
| userId |
| workerPID |



Go ahead and imagine what our workers should do.

##### Whenever a user does "SignIn":

    1. Change something in User and Session entities (e.g ` User.online = 1 ` etc)
    1. Add relation between User and Session
    3. Add a new record into ` online `

Now we able to perform queries like ` get all online users ` or ` get all online sessions ` or even ` get all online sessions by workerPID `. Aha, sounds good but we still have a big problem.

@seme1 wrote:
> the problem that may arise when one of the workers crashes.

For finding out a solution, let's imagine what happens whenever a worker does "oh noo I'm died.."?
Right! A new worker is borning. So, it's the right moment to us to clear ` online ` table from users, that actually is not online anymore, cuz previous worker has died. What our slave should do?

##### Whenever a worker is died and/or a new one is started:

    1. Ask a mysterious Master process "Hello, my Master! Give me, please, PIDs of all workers"
    2. Select all rows in the ` online ` table, that have unknown ` workerPID `. (let's call it ` dirtyOnline `)
    3. Do some additional logic, you know what you want. Just for instance: for each user in ` dirtyOnline ` check whether he has another online sessions, if no then change ` User.online = 0 ` etc
    4. Finally delete all ` dirtyOnline ` rows

This approach will also clear entire ` online ` table when you perform cold start your application.
And ` get all online session where userId X ` it's probably a good answer to your question:
> What I'm trying to achieve is to update a backend MySQL database with the number of LIVE socket sessions each user has.

Okay, now it's sounds good. Good enough? Actually **no**, I never ever tested it, and probably never will.

---

But as I see you wanna change your application. And it's good decision. Let's say now we create two SC channels ` online_users ` and ` online_anons `.
What's now?


##### Whenever a user does "SignIn":

    1. Unsubscribe user from ` online_anons `
    2. Subscribe user for ` online_users `
    3. Perform additional logic, you know, something like "Add relation between User and Session"

##### Whenever a worker is died and/or a new one is started:

    0. **Nothing**

But now we have even more questions. And those questions for @jondubois 
I'll make another issue with list of those questions soon. @seme1 please be free to add there something.

**Comment by seme1** [comment]
Thanks a lot for your comments @MegaGM . The approach of having two channels, namely `online_anons` and `online_users` is even better, since no master process is needed. It's actually what I almost settled on and I included a link in my last post to somebody who used a similar method with Socket.io. 

For me, the main remaining part now is how to obtain the number of users that are currently subscribed to a channel from the server side. There are other open issues regarding server side sub/pub but I'll include it in a separate issue.

**Comment by jondubois** [comment]
You need a database (or use the nData memory Store in SC) to keep track of users. Auth tokens should make this easier, but you can't avoid using a data store of sorts (and the solution you choose would depends on you specific needs).
